### PR TITLE
Refactor TTim to compute per log t-interval

### PR DESCRIPTION
**Pull request by dbrakenhoff**
First attempt at refactoring TTim. Changes only done for Well Element and everything needed to compute the Theis solution.

If this looks okay, I'll implement the changes for all Elements.

### CODE REVIEW

**Comment by mbakker7** [comment]
Great start!
Couple of questions and things we should still do:
1. We stick with dtype as `"D"` to specify a complex array. Is that indeed the recommended way to do this? Maybe also use `dtype` keyword here?
2. In aquifer.py we differentiate between 'lea' and 'sem'. We need to document better what the diffference is. 
3. In element.py, `potentialall` function is for testing only, I presume. Should we add an starting underscore?
4. Doc strings in element.py need to be updated to indicate correct size of returned array
5. `disvec` and `disvecinf` functions in element.py need to be modified.
6. In model.py,  in `initialize_interval` I don't quite understand why we need to sort `self.logtintervals`?
7. In model.py,  # NOTE: this check no longer allows the right boundary of the time interval to  be computed, which was allowed in the previous version of TTim. We need to discuss that.
8. 'pot = pot * aq.eigvec[t_int][layers]' is much nicer than `pot = np.sum(pot[:, np.newaxis, :, :] * aq.eigvec2[layers, :, jtime], 2)`, so I think we are on the right track!

**Comment by dbrakenhoff** [comment]
> 1. We stick with dtype as "D" to specify a complex array. Is that indeed the recommended way to do this? Maybe also use dtype keyword here?

Good idea, we can use dtype=complex instead. 

> 3. In element.py, potentialall function is for testing only, I presume. Should we add an starting underscore?

Yea this is a temporary thing, I wasn't sure yet if we still want to have a method that is maybe slightly more efficient in calculating the heads across all time-intervals. But good idea to add the underscore if it's only for testing. 

> 6. In model.py, in initialize_interval I don't quite understand why we need to sort self.logtintervals?

I thought it would be nice if logtintervals was always sorted, even after solving a separate intervals in a random order. So that's why.

> 7. In model.py, # NOTE: this check no longer allows the right boundary of the time interval to be computed, which was allowed in the previous version of TTim. We need to discuss that.

This should be very simple to add back in. The way I programmed it now was to check whether t_int was in Model.logtintervals, but we can easily add back a comparison to the actual intervals, or add a special case for the final interval. 

> 8. 'pot = pot * aq.eigvec[t_int][layers]' is much nicer than pot = np.sum(pot[:, np.newaxis, :, :] * aq.eigvec2[layers, :, jtime], 2), so I think we are on the right track!

Agreed!

**Comment by mbakker7** [comment]
Fixed storing parameters in solve, and potinflayers, unitpotentiallayers. 
Need to modify disinf and related functions in element.py
Write documentation on these functions.

File: ttim/well.py [hunk 1]
```diff
@@ -292,6 +296,18 @@ def __init__(
...
+    def initialize(self):
+        super().initialize()
+        self.parameters = {}
+        for t_int in self.model.logtintervals:
+            self.initialize_interval(t_int)
+
+    def initialize_interval(self, t_int):
+        super().initialize_interval(t_int)
+        self.parameters[t_int] = np.zeros(
+            (self.model.ngvbc, self.nparam, self.model.nppar), "D"
```
Comment (dbrakenhoff) [inline ttim/well.py:308]: dtype=complex?

File: ttim/well.py [hunk 2]
```diff
@@ -406,13 +464,19 @@ def __init__(
...
+        self.parameters = {}
+        for t_int in self.model.logtintervals:
+            self.initialize_interval(t_int)
         # Needed in solving for a unit head
         self.pc = self.aq.T[self.layers]
 
+    def initialize_interval(self, t_int):
+        super().initialize_interval(t_int)
+        self.parameters[t_int] = np.zeros(
+            (self.model.ngvbc, self.nparam, self.model.nppar), "D"
```
Comment (dbrakenhoff) [inline ttim/well.py:477]: dtype=complex

Comment (mbakker7) [reply ttim/well.py]: This is now fixed. What is still undesirable is that now all Well classes have an `initialize_interval` function. Seems repetitive. Let's discuss.

**Comment by mbakker7** [comment]
I think this PR can be closed and deleted as we have decided, for good reasons, that we are not going to do this.
The reason is that the computation of the head at a point in time may depend on many log-intervals, as the head is a superposition in time of a bunch of step responses (a new step response is computed after every change in boundary condition, e.g., a change in the discharge of a well or the head in a river).  
What we do like in the near future is better documentation of the size and shape of matrices that are computed and multiplied at different parts in the code.

**Pull request closed without merging**